aten/src/ATen/native/cuda/TensorTopK.cu

```diff
@@ -0,0 +1,338 @@
+                                                                          \
+    dim3 block(std::min(at::cuda::ATenCeilDiv(sliceSize, (int64_t) C10_WARP_SIZE)*(int64_t) C10_WARP_SIZE, (int64_t) 1024)); \
+                                                                          \
+    /* This is used as a template parameter to calculate indices. */      \
+    /* We only specialize it if all collapsed dim sizes are the */        \
+    /* same; otherwise, we use -1 which is the specialization */          \
+    /* parameter for arbitrary dimensions */                              \
+    int allDims = inputInfo.dims;                                         \
+    if (topKInfo.dims != allDims || indicesInfo.dims != allDims) {        \
+      allDims = -1;                                                       \
+    }                                                                     \
+                                                                          \
+    RUN_DIM(INDEX_T);                                                     \
+  });
+
+  // the below is safe with 0-dimensional tensors because it is based on
+  // TensorInfo which implicitly expands to 1-dimensional.
+  if (input.numel() > 0) {
+    // Based on required index size, run the algorithm with the
+    // appropriate index type
+    if (at::cuda::detail::canUse32BitIndexMath(input) &&
+        at::cuda::detail::canUse32BitIndexMath(values) &&
+        at::cuda::detail::canUse32BitIndexMath(indices)) {
+      RUN_T(uint32_t);
+    } else {
+      RUN_T(uint64_t);
+    }
+  }
+#undef RUN_T
+#undef RUN_DIM
+#undef RUN_DIR
+#undef RUN_K
+
+  // Sort the results if the user wants them sorted, since our
+  // selection routine does not ensure sorting
+  if (sorted && values.numel() > 1) {
+    // FIXME: the k/v inplace sort along slice only works for size <=
+    // 2048 at the moment
+    // Workaround:
+    int maxSliceSize;
+    if (values.element_size() >= 8) {
+      maxSliceSize = 1024;
+    } else {
+      maxSliceSize = 2048;
+    }
+    if (k <= maxSliceSize) {
+      // This avoids any memory allocations and performs all sorting
+      // work inplace along the slice
+
+      sortKeyValueInplace(values, indices, dim, largest);
+    } else {
+      // Depend upon the backup sort that returns indices, which we
+      // can use in conjunction with gather to produce the original
+      // indices.
+      // This is not the most efficient implementation, especially since
+      // there are memory allocations performed here. If the user desires
+      // greater performance, they should torch.gather() the results
+      // themselves using the reported indices, providing previously
+      // allocated tensors to receive the results.
+
+      Tensor sortedIndices = at::empty_like(indices);
+      at::native::sort_out_cuda(values, dim, largest, values, sortedIndices);
+      indices.copy_(indices.gather(dim, sortedIndices));
+    }
+  }
+  AT_CUDA_CHECK(cudaGetLastError());
```

Comment:
why do you need cuda check here? Kernel launches are checked above, and sortKeyValueInplace and other ops should have their own checks.